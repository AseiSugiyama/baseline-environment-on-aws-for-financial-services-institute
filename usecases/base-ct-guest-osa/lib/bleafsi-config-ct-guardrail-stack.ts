import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { cloudformation_include as cfn_inc } from 'aws-cdk-lib';
import { aws_config as config } from 'aws-cdk-lib';

/*
 * Control Tower ガードレール相当のConfig Rulesを設定する
 */
export class ConfigCtGuardrailStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // https://github.com/awslabs/aws-config-rules/tree/master/aws-config-conformance-packs
    new cfn_inc.CfnInclude(this, 'ConfigCtGr', {
      templateFile: 'cfn/AWS-Control-Tower-Detective-Guardrails.yaml',
    });

    // デフォルトセキュリティグループが閉じている（inbound/outboundを許可していない）ことをチェックするConfigRuleを作成
    // ConfigRule for Default Security Group is closed
    //
    // See: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#securityhub-cis-controls-4.3
    // See: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html
    new config.ManagedRule(this, 'BLEAFSIRuleDefaultSecurityGroupClosed', {
      identifier: config.ManagedRuleIdentifiers.VPC_DEFAULT_SECURITY_GROUP_CLOSED,
      ruleScope: config.RuleScope.fromResources([config.ResourceType.EC2_SECURITY_GROUP]),
      configRuleName: 'bb-default-security-group-closed',
      description:
        'Checks that the default security group of any Amazon Virtual Private Cloud (VPC) does not allow inbound or outbound traffic. The rule is non-compliant if the default security group has one or more inbound or outbound traffic.',
    });
  }
}
