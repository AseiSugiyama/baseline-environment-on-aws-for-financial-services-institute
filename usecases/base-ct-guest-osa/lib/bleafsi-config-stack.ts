import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_config as config } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

/*
 * この stack は CloudFormation stack を作成するために使用されます。
 * 詳しくは lib/bleafsi-config-stackset.ts を参照して下さい。
 */
interface ConfigStackProps extends cdk.StackProps {
  cloudTrailBucketName: string; //Log Archiveアカウントに作成したS3バケット名（cdk.jsonで設定）
}

export class ConfigStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ConfigStackProps) {
    super(scope, id, props);

    const role = new iam.Role(this, 'ConfigRole', {
      assumedBy: new iam.ServicePrincipal('config.amazonaws.com'),
      managedPolicies: [iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWS_ConfigRole')],
    });

    new config.CfnConfigurationRecorder(this, 'ConfigRecorder', {
      roleArn: role.roleArn,
      recordingGroup: {
        allSupported: true,
        includeGlobalResourceTypes: true,
      },
    });

    //create config
    new config.CfnDeliveryChannel(this, 'ConfigDeliveryChannel', {
      s3BucketName: props.cloudTrailBucketName,
    });
  }
}
