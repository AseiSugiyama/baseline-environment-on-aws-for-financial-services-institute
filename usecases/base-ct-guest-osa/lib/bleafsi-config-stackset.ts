import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';
import * as path from 'path';

/*
 * このCDKテンプレートは 'bleafsi-config-stack.ts' から生成されたCloudFormation Stackテンプレート（bleafsi-base-ct-config.yaml）を
 * 元にCloudFormation StackSetを作成し、ゲストアカウントにCloudFormation Stackをデプロイします。CloudFormation Stackテンプレートは、
 * 下記コマンドで生成して下さい。
 *
 * npx cdk synthesize --all --app "npx ts-node --prefer-ts-exts bin/bleafsi-base-generate-stack.ts" -c environment=dev --profile ct-guest > bleafsi-base-ct-config.yaml
 *
 * 直接CDKテンプレートを使ってデプロイを行わないのは、Control Tower配下の組織ではガードレールにより管理アカウント配下の
 * AWSControlTowerStackSetRole ロール以外から AWS Config をデプロイできないためです。
 *
 * このCDKテンプレートは、Control Towerの管理者アカウントで実行して下さい。ゲストアカウント上では実行できません。*/

interface ConfigStackSetProps extends cdk.StackProps {
  targetGuestAccountId: string; //guest account id to deploy
}

export class ConfigStackSet extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ConfigStackSetProps) {
    super(scope, id, props);

    const templateFilePath = path.resolve(__dirname, '../bleafsi-base-ct-config.yaml');
    const templateToString = fs.readFileSync(templateFilePath).toString();

    const regions = ['ap-northeast-3']; //Osaka region

    //create stackset
    new cdk.CfnStackSet(this, 'ConfigStackSet', {
      stackSetName: 'BLEAFSI-ConfigStackSet',
      permissionModel: 'SELF_MANAGED',
      capabilities: ['CAPABILITY_IAM'],
      administrationRoleArn: `arn:aws:iam::${cdk.Stack.of(this).account}:role/service-role/AWSControlTowerStackSetRole`,
      executionRoleName: 'AWSControlTowerExecution',
      stackInstancesGroup: [
        {
          regions: regions,
          deploymentTargets: {
            accounts: [props.targetGuestAccountId],
          },
        },
      ],
      templateBody: templateToString,
    });
  }
}
