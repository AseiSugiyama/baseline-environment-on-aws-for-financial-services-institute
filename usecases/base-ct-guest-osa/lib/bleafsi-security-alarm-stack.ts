import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_iam as iam } from 'aws-cdk-lib';
import { aws_sns as sns } from 'aws-cdk-lib';
import { aws_events as cwe } from 'aws-cdk-lib';
import { aws_events_targets as cwet } from 'aws-cdk-lib';

/*
 * SNS Topicによるアラームメール通知の設定
 */

interface SecurityAlarmStackProps extends cdk.StackProps {
  notifyEmail: string;
}

export class SecurityAlarmStack extends cdk.Stack {
  public readonly alarmTopic: sns.Topic;

  constructor(scope: Construct, id: string, props: SecurityAlarmStackProps) {
    super(scope, id, props);

    // SNS Topic for Security Alarm
    const secTopic = new sns.Topic(this, 'SecurityAlarmTopic');
    new sns.Subscription(this, 'SecurityAlarmEmail', {
      endpoint: props.notifyEmail,
      protocol: sns.SubscriptionProtocol.EMAIL,
      topic: secTopic,
    });
    this.alarmTopic = secTopic;

    // Allow to publish message from CloudWatch
    secTopic.addToResourcePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        principals: [new iam.ServicePrincipal('cloudwatch.amazonaws.com')],
        actions: ['SNS:Publish'],
        resources: [secTopic.topicArn],
      }),
    );

    // Allow to publish through ssl
    secTopic.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'AllowPublishThroughSSLOnly',
        actions: ['SNS:Publish'],
        effect: iam.Effect.DENY,
        principals: [new iam.AnyPrincipal()],
        resources: [secTopic.topicArn],
        conditions: {
          Bool: {
            'aws:SecureTransport': false,
          },
        },
      }),
    );

    // --------------- ConfigRule Compliance Change Notification -----------------
    // ConfigRule - Compliance Change
    //  See: https://docs.aws.amazon.com/config/latest/developerguide/monitor-config-with-cloudwatchevents.html
    //  See: https://aws.amazon.com/premiumsupport/knowledge-center/config-resource-non-compliant/?nc1=h_ls
    //  通知するルールを追加したい場合は、'configRuleName'配列にルール名を追加して下さい
    //  Sample Rule 'bb-default-security-group-closed' は base-ct-guest/lib/bleafsi-config-rules-stack.ts を参照
    new cwe.Rule(this, 'BLEAFSIRuleConfigRules', {
      description: 'EventBridge Event Rule to send notification on Config Rule compliance changes.',
      enabled: true,
      eventPattern: {
        source: ['aws.config'],
        detailType: ['Config Rules Compliance Change'],
        detail: {
          configRuleName: ['bb-default-security-group-closed'],
          newEvaluationResult: {
            complianceType: ['NON_COMPLIANT'],
          },
        },
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });

    // ------------------------ AWS Health Notification ---------------------------

    // AWS Health - Notify any events on AWS Health
    // See: https://aws.amazon.com/premiumsupport/knowledge-center/cloudwatch-notification-scheduled-events/?nc1=h_ls
    new cwe.Rule(this, 'BLEFSIARuleAwsHealth', {
      description: 'Notify AWS Health event',
      enabled: true,
      eventPattern: {
        source: ['aws.health'],
        detailType: ['AWS Health Event'],
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });

    // ------------ Detective guardrails from NIST standard template ----------------
    // See: https://aws.amazon.com/blogs/publicsector/automating-compliance-architecting-for-fedramp-high-and-nist-workloads-in-aws-govcloud-us/

    // Security Groups Change Notification
    // See: https://aws.amazon.com/premiumsupport/knowledge-center/monitor-security-group-changes-ec2/?nc1=h_ls
    //  from NIST template
    new cwe.Rule(this, 'BLEAFSIRuleSecurityGroupChange', {
      description: 'Notify to create, update or delete a Security Group.',
      enabled: true,
      eventPattern: {
        source: ['aws.ec2'],
        detailType: ['AWS API Call via CloudTrail'],
        detail: {
          eventSource: ['ec2.amazonaws.com'],
          eventName: [
            'AuthorizeSecurityGroupIngress',
            'AuthorizeSecurityGroupEgress',
            'RevokeSecurityGroupIngress',
            'RevokeSecurityGroupEgress',
          ],
        },
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });

    // Network ACL Change Notification
    //  from NIST template
    new cwe.Rule(this, 'BLEAFSIRuleNetworkAclChange', {
      description: 'Notify to create, update or delete a Network ACL.',
      enabled: true,
      eventPattern: {
        source: ['aws.ec2'],
        detailType: ['AWS API Call via CloudTrail'],
        detail: {
          eventSource: ['ec2.amazonaws.com'],
          eventName: [
            'CreateNetworkAcl',
            'CreateNetworkAclEntry',
            'DeleteNetworkAcl',
            'DeleteNetworkAclEntry',
            'ReplaceNetworkAclEntry',
            'ReplaceNetworkAclAssociation',
          ],
        },
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });

    // CloudTrail Change
    //  from NIST template
    /*new cwe.Rule(this, 'BLEAFSIRuleCloudTrailChange', {
      description: 'Notify to change on CloudTrail log configuration',
      enabled: true,
      eventPattern: {
        detailType: ['AWS API Call via CloudTrail'],
        detail: {
          eventSource: ['cloudtrail.amazonaws.com'],
          eventName: ['StopLogging', 'DeleteTrail', 'UpdateTrail'],
        },
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });*/
    //CloudTrailマルチリージョン証跡により、大阪リージョンではなく東京リージョンでCloudTrail証跡が取られるため、CloudTrail自体のイベントによるアラート、
    //CW Logsのアラートは大阪リージョンでは実装しない

    // ------------------- Other security services integration ----------------------

    // SecurityHub - Imported
    //   Security Hub automatically sends all new findings and all updates to existing findings to EventBridge as Security Hub Findings - Imported events.
    //   See: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cwe-integration-types.html
    //
    //   Security Hub Finding format
    //   See: https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-findings-format.html
    new cwe.Rule(this, 'BLEAFSIRuleSecurityHub', {
      description: 'EventBridge Event Rule to send notification on SecurityHub all new findings and all updates.',
      enabled: true,
      eventPattern: {
        source: ['aws.securityhub'],
        detailType: ['Security Hub Findings - Imported'],
        detail: {
          findings: {
            Severity: {
              Label: ['CRITICAL', 'HIGH'],
            },
            Compliance: {
              Status: ['FAILED'],
            },
            Workflow: {
              Status: ['NEW', 'NOTIFIED'],
            },
            RecordState: ['ACTIVE'],
          },
        },
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });

    // GuardDutyFindings
    //   Will alert for any Medium to High finding.
    //   See: https://docs.aws.amazon.com/guardduty/latest/ug/guardduty_findings_cloudwatch.html
    new cwe.Rule(this, 'BLEAFSIRuleGuardDuty', {
      description: 'EventBridge Event Rule to send notification on GuardDuty findings.',
      enabled: true,
      eventPattern: {
        source: ['aws.guardduty'],
        detailType: ['GuardDuty Finding'],
        detail: {
          severity: [
            4, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6,
            6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8,
            8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9,
          ],
        },
      },
      targets: [new cwet.SnsTopic(secTopic)],
    });
  }
}
