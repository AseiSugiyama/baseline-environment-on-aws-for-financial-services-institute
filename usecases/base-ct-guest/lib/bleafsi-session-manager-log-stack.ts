import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

/*
 * Session Managerログ保管用のS3バケットの作成
 * S3バケットはゲストアカウント上に作成される
 */

export class SessionManagerLogStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: cdk.StackProps) {
    super(scope, id, props);

    // Archive Bucket for CloudTrail
    const archiveLogsBucket = new s3.Bucket(this, 'ArchiveLogsBucket', {
      accessControl: s3.BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      enforceSSL: true,
      lifecycleRules: [
        {
          enabled: true,
          expiration: cdk.Duration.days(2555),
          transitions: [
            {
              transitionAfter: cdk.Duration.days(90),
              storageClass: s3.StorageClass.GLACIER,
            },
          ],
        },
      ],
    });
    //protection from deleting objects
    archiveLogsBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'Restrict Delete* Actions',
        effect: iam.Effect.DENY,
        actions: ['s3:DeleteObject'],
        principals: [new iam.AnyPrincipal()],
        resources: [archiveLogsBucket.arnForObjects('*')],
      }),
    );
    // ## Create a Managed Policy for Session Manager S3 Write
    const inlineIamPolicy1 = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['s3:PubObjectAcl', 's3:PutObject'],
    });
    const inlineIamPolicy2 = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['s3:GetEncryptionConfiguration'],
    });

    new iam.ManagedPolicy(this, `WritePolicy`, {
      statements: [inlineIamPolicy1, inlineIamPolicy2],
    });

    // ## Bucket for Session Manager Log
    const sessionManageLogBucket = new s3.Bucket(this, 'SessionManagerLogsBucket', {
      accessControl: s3.BucketAccessControl.PRIVATE,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      encryption: s3.BucketEncryption.S3_MANAGED,
      enforceSSL: true,
      serverAccessLogsBucket: archiveLogsBucket,
    });
    //protection from deleting objects
    sessionManageLogBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'Restrict Delete* Actions',
        effect: iam.Effect.DENY,
        actions: ['s3:DeleteObject'],
        principals: [new iam.AnyPrincipal()],
        resources: [sessionManageLogBucket.arnForObjects('*')],
      }),
    );

    //add bucket resource to role's inline policy
    inlineIamPolicy1.addResources(`${sessionManageLogBucket.bucketArn}/AWSLogs/*`);
    inlineIamPolicy2.addResources(`${sessionManageLogBucket.bucketArn}`);

    //## CFn output
    new cdk.CfnOutput(this, 'SSM Session Manager Log Bucket', {
      value: sessionManageLogBucket.bucketName,
      description: 'Bucket for SSM Session Manager Log Bucket',
    });
  }
}
