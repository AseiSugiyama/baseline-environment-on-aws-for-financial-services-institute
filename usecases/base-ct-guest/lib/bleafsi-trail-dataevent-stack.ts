import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudtrail as trail } from 'aws-cdk-lib';
import { aws_logs as cwl } from 'aws-cdk-lib';
import { aws_kms as kms } from 'aws-cdk-lib';
import { S3EventSelector } from 'aws-cdk-lib/aws-cloudtrail';

/*
 * CloudTrail S3データイベントの有効化
 * ログはLog Archiveアカウントの集約バケットに保管される
 */

interface TrailStackProps extends cdk.StackProps {
  cloudTrailBucketName: string; //Log Archiveアカウントに作成したS3バケット名（cdk.jsonで設定）
  targetBuckets: string[]; //CloudTrailデータイベント取得の対象となるS3バケット名を指定（cdk.jsonで指定）
  controlTowerKMSKeyArn: string; //CloudTrail 暗号化キー（ControlTowerの暗号化キーを使用. cdk.jsonで指定）
}

export class TrailDataEventStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: TrailStackProps) {
    super(scope, id, props);

    // Enryption key for CloudTrail. Use KMS key in Control Tower
    const ctKmsKey = kms.Key.fromKeyArn(this, 'ct-kmskey', props.controlTowerKMSKeyArn);

    // CloudWatch Logs Group for CloudTrail data logging
    const cloudTrailLogGroup = new cwl.LogGroup(this, 'CloudTrailLogGroup', {
      retention: cwl.RetentionDays.TWO_WEEKS,
    });

    //find a bucket for cloudtrail
    const cloudTrailBucket = s3.Bucket.fromBucketName(this, 'cloudTrailBucket', props.cloudTrailBucketName);

    // generate resources arn for field selectors
    const s3eventSelectors: S3EventSelector[] = [];
    for (const bucketName of props.targetBuckets) {
      s3eventSelectors.push({ bucket: s3.Bucket.fromBucketName(this, bucketName, bucketName) });
    }

    // Create CloudTrail for data logging
    const cloudtrail = new trail.Trail(this, 'CloudTrail', {
      trailName: 's3-dataevent-trail',
      bucket: cloudTrailBucket,
      enableFileValidation: true,
      includeGlobalServiceEvents: true,
      cloudWatchLogGroup: cloudTrailLogGroup,
      encryptionKey: ctKmsKey,
      sendToCloudWatchLogs: true,
    });

    cloudtrail.addS3EventSelector(s3eventSelectors, {
      includeManagementEvents: false,
      readWriteType: trail.ReadWriteType.ALL,
    });
  }
}
