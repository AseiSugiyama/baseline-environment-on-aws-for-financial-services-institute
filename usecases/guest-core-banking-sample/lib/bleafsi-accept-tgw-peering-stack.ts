import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { custom_resources as cr } from 'aws-cdk-lib';
import { CrossRegionSsmParamName } from './shared/bleafsi-constants';
import { CrossRegionSsmParam } from './shared/bleafsi-cross-region-ssm-param-stack';

interface AcceptTgwPeeringStackProps extends cdk.StackProps {
  pjPrefix: string;
  envName: string;
}

export class AcceptTgwPeeringStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: AcceptTgwPeeringStackProps) {
    super(scope, id, props);

    const { pjPrefix, envName } = props;

    const crossRegionSsmParam = new CrossRegionSsmParam(this, `${pjPrefix}-crossRegionSsmParam`, {
      baseRegion: cdk.Stack.of(this).region,
      envName,
    });
    const tgwPeeringAttachmentId = crossRegionSsmParam.get(CrossRegionSsmParamName.TGW_PEERING_ATTACHMENT_ID);

    // Accept Transit Gateway Peering Attachment via API
    const crStackId = `${pjPrefix}-AcceptTgwPeeringAttachment`;
    new cr.AwsCustomResource(this, crStackId, {
      onUpdate: {
        service: 'EC2',
        action: 'acceptTransitGatewayPeeringAttachment',
        parameters: {
          TransitGatewayAttachmentId: tgwPeeringAttachmentId,
          DryRun: false,
        },
        region: cdk.Stack.of(this).region,
        physicalResourceId: cr.PhysicalResourceId.of(crStackId),
      },
      onDelete: {
        service: 'EC2',
        action: 'deleteTransitGatewayPeeringAttachment',
        parameters: {
          TransitGatewayAttachmentId: tgwPeeringAttachmentId,
          DryRun: false,
        },
        region: cdk.Stack.of(this).region,
        physicalResourceId: cr.PhysicalResourceId.of(crStackId),
      },
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
        resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
