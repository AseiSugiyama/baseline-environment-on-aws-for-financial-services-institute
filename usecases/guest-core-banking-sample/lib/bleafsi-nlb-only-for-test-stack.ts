import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_ec2 as ec2 } from 'aws-cdk-lib';
import { aws_elasticloadbalancingv2 as elbv2 } from 'aws-cdk-lib';
import { aws_elasticloadbalancingv2_targets as elbv2_targets } from 'aws-cdk-lib';

interface NlbOnlyForTestContextProps extends cdk.StackProps {
  pjPrefix: string;
  myVpc: ec2.Vpc;
  targetAlb: elbv2.ApplicationLoadBalancer;
}

export class NlbOnlyForTestStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: NlbOnlyForTestContextProps) {
    super(scope, id, props);

    const { pjPrefix, myVpc, targetAlb } = props;

    // Network Load Balancer for test
    const lb = new elbv2.NetworkLoadBalancer(this, `${pjPrefix}-NlbOnlyForTest`, {
      vpc: myVpc,
      internetFacing: true,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
    });
    const listener = lb.addListener(`${pjPrefix}-TestNlbListener`, {
      port: 80,
    });
    listener.addTargets(`${pjPrefix}-TargetAlb`, {
      port: 80,
      targets: [new elbv2_targets.AlbArnTarget(targetAlb.loadBalancerArn, 80)],
    });

    // You can access this URL over the Internet.
    new cdk.CfnOutput(this, 'TestNlbUrl', {
      value: `http://${lb.loadBalancerDnsName}`,
    });
  }
}
