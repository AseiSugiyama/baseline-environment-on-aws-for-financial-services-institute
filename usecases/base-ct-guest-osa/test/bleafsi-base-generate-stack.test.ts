import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Annotations, Match } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks } from 'cdk-nag';
import { NagSuppressions } from 'cdk-nag';

import { ConfigStack } from '../lib/bleafsi-config-stack';

const procEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new cdk.App();
const pjPrefix = 'BLEA-FSI-BASE';

let stack: cdk.Stack;

//jest snapshot check
describe(`${pjPrefix} ControlTower Stacks`, () => {
  test('GuestOsa3Account Stacks', () => {
    stack = new ConfigStack(app, `${pjPrefix}-config-stack`, {
      cloudTrailBucketName: 'dummy bucket', //S3 bucket name created by cdk template in Log Archive account
      env: procEnv,
    });

    // test with snapshot
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});

//cdk-nag check
describe(`${pjPrefix} cdk-nag AwsSolutions Pack`, () => {
  beforeAll(() => {
    //cdk-nag suppression
    // IAM4: The IAM user, role, or group uses AWS managed policies.
    NagSuppressions.addResourceSuppressionsByPath(stack, '/BLEA-FSI-BASE-config-stack/ConfigRole/Resource', [
      {
        id: 'AwsSolutions-IAM4',
        reason: 'this stack is used for administrative task. this role uses service-linked role',
      },
    ]);

    //cdk-nag check
    cdk.Aspects.of(stack).add(new AwsSolutionsChecks());
  });

  test('No unsupressed Errors', () => {
    const errors1 = Annotations.fromStack(stack).findError('*', Match.stringLikeRegexp('AwsSolutions-.*'));
    try {
      expect(errors1).toHaveLength(0);
      console.log('cdk-nag: no errors for stack ' + stack.stackName);
    } catch (e) {
      console.error(errors1);
      throw e;
    }
  });
});
