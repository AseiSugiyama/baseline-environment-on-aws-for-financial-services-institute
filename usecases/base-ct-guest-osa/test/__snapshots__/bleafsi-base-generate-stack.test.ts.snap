// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BLEA-FSI-BASE ControlTower Stacks GuestOsa3Account Stacks 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ConfigDeliveryChannel": Object {
      "Properties": Object {
        "S3BucketName": "dummy bucket",
      },
      "Type": "AWS::Config::DeliveryChannel",
    },
    "ConfigRecorder": Object {
      "Properties": Object {
        "RecordingGroup": Object {
          "AllSupported": true,
          "IncludeGlobalResourceTypes": true,
        },
        "RoleARN": Object {
          "Fn::GetAtt": Array [
            "ConfigRoleF94F46B6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Config::ConfigurationRecorder",
    },
    "ConfigRoleF94F46B6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "config.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWS_ConfigRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
